File: apps/mobile/targets/keyboard/SpokenEntities.swift
import Foundation

enum SpokenEntities {
    struct Match {
        let consumed: Int
        let text: String
    }

    private struct Atom {
        let value: Int
        let scale: Int

        init(_ value: Int, scale: Int = 1) {
            self.value = value
            self.scale = scale
        }
    }

    private static let numberValues: [String: [String: Int]] = [
        "en": [
            "zero": 0, "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
            "six": 6, "seven": 7, "eight": 8, "nine": 9, "ten": 10,
            "eleven": 11, "twelve": 12, "thirteen": 13, "fourteen": 14,
            "fifteen": 15, "sixteen": 16, "seventeen": 17, "eighteen": 18,
            "nineteen": 19, "twenty": 20, "thirty": 30, "forty": 40,
            "fifty": 50, "sixty": 60, "seventy": 70, "eighty": 80, "ninety": 90
        ],
        "es": [
            "cero": 0, "uno": 1, "un": 1, "una": 1, "dos": 2, "tres": 3,
            "cuatro": 4, "cinco": 5, "seis": 6, "siete": 7, "ocho": 8,
            "nueve": 9, "diez": 10, "once": 11, "doce": 12, "trece": 13,
            "catorce": 14, "quince": 15, "dieciséis": 16, "diecisiete": 17,
            "dieciocho": 18, "diecinueve": 19, "veinte": 20, "veintiuno": 21,
            "veintiún": 21, "veintiuna": 21, "veintidós": 22, "veintitrés": 23,
            "veinticuatro": 24, "veinticinco": 25, "veintiséis": 26,
            "veintisiete": 27, "veintiocho": 28, "veintinueve": 29,
            "treinta": 30, "cuarenta": 40, "cincuenta": 50, "sesenta": 60,
            "setenta": 70, "ochenta": 80, "noventa": 90
        ],
        "pt": [
            "zero": 0, "um": 1, "uma": 1, "dois": 2, "duas": 2, "três": 3,
            "quatro": 4, "cinco": 5, "seis": 6, "sete": 7, "oito": 8,
            "nove": 9, "dez": 10, "onze": 11, "doze": 12, "treze": 13,
            "catorze": 14, "quatorze": 14, "quinze": 15, "dezesseis": 16,
            "dezasseis": 16, "dezessete": 17, "dezassete": 17, "dezoito": 18,
            "dezenove": 19, "vinte": 20, "trinta": 30, "quarenta": 40,
            "cinquenta": 50, "sessenta": 60, "setenta": 70, "oitenta": 80,
            "noventa": 90
        ]
    ]

    private static let months: [String: [String]] = [
        "en": [
            "january", "february", "march", "april", "may", "june",
            "july", "august", "september", "october", "november", "december"
        ],
        "es": [
            "enero", "febrero", "marzo", "abril", "mayo", "junio",
            "julio", "agosto", "septiembre", "octubre", "noviembre", "diciembre"
        ],
        "pt": [
            "janeiro", "fevereiro", "março", "abril", "maio", "junho",
            "julho", "agosto", "setembro", "outubro", "novembro", "dezembro"
        ]
    ]

    private static func isConnector(_ token: String, language: String) -> Bool {
        token == ["en": "and", "es": "y", "pt": "e"][language]
    }

    private static func atom(_ token: String, language: String) -> Atom? {
        if let value = Int(token), value >= 0 { return Atom(value) }
        if ["hundred", "cien", "ciento", "cem", "cento"].contains(token) {
            return Atom(0, scale: 100)
        }
        if token == "thousand" || token == "mil" { return Atom(0, scale: 1_000) }
        return numberValues[language]?[token].map { Atom($0) }
    }

    private static func parseNumber(_ tokens: ArraySlice<String>, language: String) -> Int? {
        guard !tokens.isEmpty else { return nil }
        var total = 0
        var current = 0
        var sawNumber = false
        for (offset, token) in tokens.enumerated() {
            if isConnector(token, language: language) {
                guard sawNumber, offset + 1 < tokens.count else { return nil }
                continue
            }
            guard let atom = atom(token, language: language) else { return nil }
            guard atom.scale == 1 || sawNumber else { return nil }
            sawNumber = true
            switch atom.scale {
            case 100:
                current = max(current, 1) * 100
            case 1_000:
                total += max(current, 1) * 1_000
                current = 0
            default:
                current += atom.value
            }
        }
        return sawNumber ? total + current : nil
    }

    private static func numberPrefixLength(
        _ tokens: [String],
        start: Int,
        language: String
    ) -> Int {
        var end = start
        while end < tokens.count, end - start < 10 {
            if atom(tokens[end], language: language) == nil,
               !isConnector(tokens[end], language: language) {
                break
            }
            end += 1
        }
        while end > start, isConnector(tokens[end - 1], language: language) { end -= 1 }
        return end - start
    }

    private static func month(_ token: String, language: String) -> Int? {
        guard let index = months[language]?.firstIndex(of: token) else { return nil }
        return index + 1
    }

    private static func date(_ tokens: [String], start: Int, language: String) -> Match? {
        if language == "en" {
            guard start < tokens.count,
                  let month = month(tokens[start], language: language) else { return nil }
            let length = numberPrefixLength(tokens, start: start + 1, language: language)
            guard length >= 2 else { return nil }
            for split in stride(from: length - 1, through: 1, by: -1) {
                let day = parseNumber(tokens[(start + 1)..<(start + 1 + split)], language: language)
                let year = parseNumber(
                    tokens[(start + 1 + split)..<(start + 1 + length)],
                    language: language
                )
                if let day, let year, (1...31).contains(day), (1_000...9_999).contains(year) {
                    return Match(
                        consumed: 1 + length,
                        text: String(format: "%04d-%02d-%02d", year, month, day)
                    )
                }
            }
            return nil
        }

        let limit = min(tokens.count, start + 6)
        guard start + 1 < limit,
              let firstDe = (start + 1..<limit).first(where: { tokens[$0] == "de" }),
              let day = parseNumber(tokens[start..<firstDe], language: language),
              let month = tokens.indices.contains(firstDe + 1)
                ? month(tokens[firstDe + 1], language: language) : nil,
              tokens.indices.contains(firstDe + 2), tokens[firstDe + 2] == "de" else {
            return nil
        }
        let yearStart = firstDe + 3
        let yearLength = numberPrefixLength(tokens, start: yearStart, language: language)
        guard yearLength > 0,
              let year = parseNumber(
                tokens[yearStart..<(yearStart + yearLength)],
                language: language
              ),
              (1...31).contains(day), (1_000...9_999).contains(year) else { return nil }
        return Match(
            consumed: yearStart + yearLength - start,
            text: String(format: "%04d-%02d-%02d", year, month, day)
        )
    }

    private static func time(_ tokens: [String], start: Int, language: String) -> Match? {
        let limit = min(tokens.count, start + 7)
        guard start + 2 < limit,
              let marker = (start + 2..<limit).first(where: {
                tokens[$0] == "am" || tokens[$0] == "pm"
              }) else { return nil }
        for split in start + 1..<marker {
            let minuteStart = isConnector(tokens[split], language: language) ? split + 1 : split
            guard minuteStart < marker else { continue }
            let hour = parseNumber(tokens[start..<split], language: language)
            let minute = parseNumber(tokens[minuteStart..<marker], language: language)
            if let hour, let minute, (1...12).contains(hour), (0...59).contains(minute) {
                return Match(
                    consumed: marker - start + 1,
                    text: String(
                        format: "%d:%02d %@",
                        hour,
                        minute,
                        tokens[marker].uppercased()
                    )
                )
            }
        }
        return nil
    }

    private static func currency(_ token: String, language: String) -> String? {
        if token == "euro" || token == "euros" { return "€" }
        if language == "en", ["pound", "pounds"].contains(token) { return "£" }
        if language == "pt", ["real", "reais"].contains(token) { return "R$" }
        if ["dólar", "dólares", "dolar", "dolares"].contains(token) {
            return language == "pt" ? "US$" : "$"
        }
        if language == "en", ["dollar", "dollars"].contains(token) { return "$" }
        if language == "es", ["peso", "pesos"].contains(token) { return "$" }
        return nil
    }

    static func match(_ tokens: [String], start: Int, language: String) -> Match? {
        if let date = date(tokens, start: start, language: language) { return date }
        if let time = time(tokens, start: start, language: language) { return time }
        let length = numberPrefixLength(tokens, start: start, language: language)
        guard length > 0,
              let value = parseNumber(tokens[start..<(start + length)], language: language) else {
            return nil
        }
        if start + length < tokens.count,
           let symbol = currency(tokens[start + length], language: language) {
            return Match(consumed: length + 1, text: "\(symbol)\(value)")
        }
        return Match(consumed: length, text: "\(value)")
    }
}

File: apps/mobile/native/android/com/j11/looper/mobile/SpokenEntities.kt
package com.j11.looper.mobile

internal object SpokenEntities {
    data class Match(val consumed: Int, val text: String)

    private val numberValues = mapOf(
        "en" to mapOf(
            "zero" to 0, "one" to 1, "two" to 2, "three" to 3, "four" to 4,
            "five" to 5, "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9,
            "ten" to 10, "eleven" to 11, "twelve" to 12, "thirteen" to 13,
            "fourteen" to 14, "fifteen" to 15, "sixteen" to 16,
            "seventeen" to 17, "eighteen" to 18, "nineteen" to 19,
            "twenty" to 20, "thirty" to 30, "forty" to 40, "fifty" to 50,
            "sixty" to 60, "seventy" to 70, "eighty" to 80, "ninety" to 90,
        ),
        "es" to mapOf(
            "cero" to 0, "uno" to 1, "un" to 1, "una" to 1, "dos" to 2,
            "tres" to 3, "cuatro" to 4, "cinco" to 5, "seis" to 6,
            "siete" to 7, "ocho" to 8, "nueve" to 9, "diez" to 10,
            "once" to 11, "doce" to 12, "trece" to 13, "catorce" to 14,
            "quince" to 15, "dieciséis" to 16, "diecisiete" to 17,
            "dieciocho" to 18, "diecinueve" to 19, "veinte" to 20,
            "veintiuno" to 21, "veintiún" to 21, "veintiuna" to 21,
            "veintidós" to 22, "veintitrés" to 23, "veinticuatro" to 24,
            "veinticinco" to 25, "veintiséis" to 26, "veintisiete" to 27,
            "veintiocho" to 28, "veintinueve" to 29, "treinta" to 30,
            "cuarenta" to 40, "cincuenta" to 50, "sesenta" to 60,
            "setenta" to 70, "ochenta" to 80, "noventa" to 90,
        ),
        "pt" to mapOf(
            "zero" to 0, "um" to 1, "uma" to 1, "dois" to 2, "duas" to 2,
            "três" to 3, "quatro" to 4, "cinco" to 5, "seis" to 6,
            "sete" to 7, "oito" to 8, "nove" to 9, "dez" to 10,
            "onze" to 11, "doze" to 12, "treze" to 13, "catorze" to 14,
            "quatorze" to 14, "quinze" to 15, "dezesseis" to 16,
            "dezasseis" to 16, "dezessete" to 17, "dezassete" to 17,
            "dezoito" to 18, "dezenove" to 19, "vinte" to 20,
            "trinta" to 30, "quarenta" to 40, "cinquenta" to 50,
            "sessenta" to 60, "setenta" to 70, "oitenta" to 80,
            "noventa" to 90,
        ),
    )

    private val months = mapOf(
        "en" to listOf(
            "january", "february", "march", "april", "may", "june",
            "july", "august", "september", "october", "november", "december",
        ),
        "es" to listOf(
            "enero", "febrero", "marzo", "abril", "mayo", "junio",
            "julio", "agosto", "septiembre", "octubre", "noviembre", "diciembre",
        ),
        "pt" to listOf(
            "janeiro", "fevereiro", "março", "abril", "maio", "junho",
            "julho", "agosto", "setembro", "outubro", "novembro", "dezembro",
        ),
    )

    private data class Atom(val value: Int, val scale: Int = 1)

    private fun connector(token: String, language: String): Boolean =
        token == mapOf("en" to "and", "es" to "y", "pt" to "e")[language]

    private fun atom(token: String, language: String): Atom? {
        token.toIntOrNull()?.takeIf { it >= 0 }?.let { return Atom(it) }
        if (token in setOf("hundred", "cien", "ciento", "cem", "cento")) {
            return Atom(0, 100)
        }
        if (token == "thousand" || token == "mil") return Atom(0, 1_000)
        return numberValues[language]?.get(token)?.let(::Atom)
    }

    private fun parseNumber(tokens: List<String>, language: String): Int? {
        if (tokens.isEmpty()) return null
        var total = 0
        var current = 0
        var sawNumber = false
        tokens.forEachIndexed { index, token ->
            if (connector(token, language)) {
                if (!sawNumber || index + 1 == tokens.size) return null
                return@forEachIndexed
            }
            val atom = atom(token, language) ?: return null
            if (atom.scale > 1 && !sawNumber) return null
            sawNumber = true
            when (atom.scale) {
                100 -> current = maxOf(current, 1) * 100
                1_000 -> {
                    total += maxOf(current, 1) * 1_000
                    current = 0
                }
                else -> current += atom.value
            }
        }
        return if (sawNumber) total + current else null
    }

    private fun numberPrefixLength(tokens: List<String>, start: Int, language: String): Int {
        var end = start
        while (end < tokens.size && end - start < 10) {
            if (atom(tokens[end], language) == null && !connector(tokens[end], language)) break
            end += 1
        }
        while (end > start && connector(tokens[end - 1], language)) end -= 1
        return end - start
    }

    private fun month(token: String, language: String): Int? {
        val index = months[language]?.indexOf(token) ?: return null
        return (index + 1).takeIf { index >= 0 }
    }

    private fun date(tokens: List<String>, start: Int, language: String): Match? {
        if (language == "en") {
            val month = tokens.getOrNull(start)?.let { month(it, language) } ?: return null
            val length = numberPrefixLength(tokens, start + 1, language)
            for (split in length - 1 downTo 1) {
                val day = parseNumber(tokens.subList(start + 1, start + 1 + split), language)
                val year = parseNumber(tokens.subList(start + 1 + split, start + 1 + length), language)
                if (day != null && year != null && day in 1..31 && year in 1_000..9_999) {
                    return Match(1 + length, "%04d-%02d-%02d".format(year, month, day))
                }
            }
            return null
        }

        val firstDe = (start + 1 until minOf(tokens.size, start + 6))
            .firstOrNull { tokens[it] == "de" } ?: return null
        val day = parseNumber(tokens.subList(start, firstDe), language) ?: return null
        val month = tokens.getOrNull(firstDe + 1)?.let { month(it, language) } ?: return null
        if (tokens.getOrNull(firstDe + 2) != "de") return null
        val yearStart = firstDe + 3
        val yearLength = numberPrefixLength(tokens, yearStart, language)
        val year = parseNumber(tokens.subList(yearStart, yearStart + yearLength), language) ?: return null
        if (day !in 1..31 || year !in 1_000..9_999) return null
        return Match(yearStart + yearLength - start, "%04d-%02d-%02d".format(year, month, day))
    }

    private fun time(tokens: List<String>, start: Int, language: String): Match? {
        val marker = (start + 2 until minOf(tokens.size, start + 7))
            .firstOrNull { tokens[it] == "am" || tokens[it] == "pm" } ?: return null
        for (split in start + 1 until marker) {
            val minuteStart = if (connector(tokens[split], language)) split + 1 else split
            val hour = parseNumber(tokens.subList(start, split), language)
            val minute = parseNumber(tokens.subList(minuteStart, marker), language)
            if (hour != null && minute != null && hour in 1..12 && minute in 0..59) {
                return Match(
                    marker - start + 1,
                    "%d:%02d %s".format(hour, minute, tokens[marker].uppercase()),
                )
            }
        }
        return null
    }

    private fun currency(token: String, language: String): String? = when {
        token == "euro" || token == "euros" -> "€"
        language == "en" && token in setOf("pound", "pounds") -> "£"
        language == "pt" && token in setOf("real", "reais") -> "R\$"
        token in setOf("dólar", "dólares", "dolar", "dolares") ->
            if (language == "pt") "US\$" else "\$"
        language == "en" && token in setOf("dollar", "dollars") -> "\$"
        language == "es" && token in setOf("peso", "pesos") -> "\$"
        else -> null
    }

    fun match(tokens: List<String>, start: Int, language: String): Match? {
        date(tokens, start, language)?.let { return it }
        time(tokens, start, language)?.let { return it }
        val length = numberPrefixLength(tokens, start, language)
        if (length == 0) return null
        val value = parseNumber(tokens.subList(start, start + length), language) ?: return null
        val symbol = tokens.getOrNull(start + length)?.let { currency(it, language) }
        return if (symbol == null) Match(length, value.toString()) else Match(length + 1, "$symbol$value")
    }
}
